### ISSUE TITLE
staradmin unload DbException

### ISSUE THREAD

**Issue by ivarbergman** [Starcounter/Starcounter#2485]
Starcounter build:

    Starcounter-DevelopNightlyBuilds-2.0.2954 (Dec 22, 2014 1:03:49 AM)

Hi! 
I'm trying to unload a database before a starcounter upgrade following these [instructions](https://github.com/Starcounter/Starcounter/wiki/Reloading-database-between-Starcounter-versions)

But I get the a DbException while doing a staradmin unload on the database named "demo7". Any ideas?

    C:\Program Files\Starcounter>staradmin --d=demo7 unload

    Unloading demo7 (unloading ..)
    Unload started at 09:39:28.6620036
    
    [09:39:29, Starcounter (demo7)]
    Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entrypoint method raised an exception. System.InvalidOperationException: Sequence contains no elements
       at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
       at Starcounter.Reload.GetPropertyName(Column col) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Starcounter\Mock\Reload.cs:line 39   at Starcounter.Reload.Unload(String fileName, UInt64 shiftId, Boolean unloadAll) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Starcounter\Mock\Reload.cs:line 71
       at UnloadDatabase.Program.Main(String[] args) in c:\Program Files\Starcounter\2.0.2954.3\cli-apps\UnloadDatabase.cs:line 24Version: 2.0.2954.3.Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4259.
       at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application) inc:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 436
       at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 193
       at Starcounter.Hosting.Package.Process() in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 123
       at Starcounter.Hosting.Package.Process(IntPtr hPackage) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 54
       at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 66
    ErrorCode=4259
    HResult=-2146233088
    HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4259
    ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
       at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
       at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
       at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
       at Starcounter.Db.ImplicitScope(Action action, Int32 maxRetries) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Starcounter\Db.cs:line 356
    
       at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Hosting\Starcount
    er.Hosting\Package.cs:line 419
    HResult=-2146232828
    ---> System.InvalidOperationException: Sequence contains no elements
       at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
       at Starcounter.Reload.GetPropertyName(Column col) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Starcounter\Mock\Reload.cs:line 39
       at Starcounter.Reload.Unload(String fileName, UInt64 shiftId, Boolean unloadAll) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2954.3\Level1\src\Starcounte
    r\Mock\Reload.cs:line 71
       at UnloadDatabase.Program.Main(String[] args) in c:\Program Files\Starcounter\2.0.2954.3\cli-apps\UnloadDatabase.cs:line 24
    HResult=-2146233079

**Comment by dan31** [comment]
Hi @ivarbergman, thanks for reporting the issue!
@k-rus, do you observe any particular fault pattern here? If no, then a full set of image files + an app is needed from @ivarbergman.

**Comment by k-rus** [comment]
Thank you for reporting the issue!
We need to reproduce it locally if possible to see what the problem is.
If the database schema is simple, can you post here?

**Comment by ivarbergman** [comment]
Hi! The error happens when unloading a database that contains our complete Tunity application. (Ie complex schema.) 

Would a raw copy of the database help you investigate further? I made a zip-file containing a copy of the database files and our application. You can download it here:

http://tunitydemo7.cloudapp.net/Issue2485.zip

Or do you need a simpler application that reproduces the error?

**Comment by k-rus** [comment]
Simpler application will be of course helpful. Do you provide only binaries for the application or source code? The ideal case for me is the source code and the database and instruction how build and run the application.

**Comment by ivarbergman** [comment]
Hi! I will talk with my colleagues to see if we can invite you to our gir repository.

Meanwhile..
I tried "staradmin unload"  on a database running the microLauncher app. This also resulted in an exception. (But not exactly the same exception as my first post in this issue.) Is this a separate issue?

1 - Fresh install of latest Starcounter Polyjuice version.  Starcounter-DevelopNightlyBuilds-2.0.2954
2 - Clone https://github.com/Polyjuice/microLauncher
3 - Configure Working Directory path to Client-folder 
4 - Start the project in VS
5 - Access http://localhost:8080/index
6 - staradmin unload

    C:\Program Files\Starcounter\2.0.2991.3>staradmin unload
    Unloading Default (unloading ..)
    
    [18:23:58, Starcounter (default)]
    Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entry point method raised an exception. System.NullReferenceException: Object reference not set to an instance of an object.
       at Starcounter.Binding.BindingBuilder.VerifyProperty(PropertyInfo propertyInfo, Type returnType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 967
       at Starcounter.Binding.BindingBuilder.GeneratePropertyBindingDefault(PropertyDef propertyDef, Type bindingBaseType, String methodName, Type returnType, TypenullableReturnType, Type thisType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 783
       at Starcounter.Binding.BindingBuilder.CreateBooleanPropertyBinding(PropertyDef propertyDef, Type thisType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 277
       at Starcounter.Binding.BindingBuilder.BuildPropertyBindingList(TypeBinding typeBinding, Type type) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 195
       at Starcounter.Binding.BindingBuilder.CreateTypeBinding() in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 151
       at Starcounter.Binding.Bindings.BuildTypeBindingFromTypeDef(TypeDef typeDef)in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\Bindings.cs:line 255
       at Starcounter.Query.Execution.RowTypeBinding.AddTypeBinding(String typeName) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\Execution\Row\RowTypeBinding.cs:line 122
       at Starcounter.Query.Sql.Creator.CreateInitialRowTypeBinding(Term typeListTerm) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\SQL\_Creator.cs:line 392
       at Starcounter.Query.Sql.Creator.CreateRowTypeBinding(Term typeDefTerm, VariableArray varArray) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\SQL\_Creator.cs:line 356
       at Starcounter.Query.QueryPreparation.PrepareOrExecuteQuery[T](String query,Boolean slowSql, Object[] values) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\QueryPreparation.cs:line 73
       at Starcounter.Query.Sql.SqlEnumCache.CacheOrExecuteEnumerator[T](String query, Boolean slowSQL, Object[] values) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\SQL\_SqlEnumCache.cs:line 123
       at Starcounter.Db.FullQueryProcess[T](String query, Boolean slowSQL, Object[] values) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Mock\Db.cs:line 126
       at Starcounter.Reload.Unload(String fileName, UInt64 shiftId, Boolean unloadAll) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Mock\Reload.cs:line 80
       at UnloadDatabase.Program.Main(String[] args) in c:\Program Files\Starcounter\2.0.2991.3\cli-apps\UnloadDatabase.cs:line 24Version: 2.0.2991.3.Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4259.
       at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application) inc:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 428
       at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 193
       at Starcounter.Hosting.Package.Process() in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 125
       at Starcounter.Hosting.Package.Process(IntPtr hPackage) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 56
       at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 66
    ErrorCode=4259
    HResult=-2146233088
    HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4259
    ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
       at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
       at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
       at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
       at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application) in
    c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 414
    HResult=-2146232828
    ---> System.NullReferenceException: Object reference not set to an instance of an object.
       at Starcounter.Binding.BindingBuilder.VerifyProperty(PropertyInfo propertyInfo, Type returnType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 967
       at Starcounter.Binding.BindingBuilder.GeneratePropertyBindingDefault(PropertyDef propertyDef, Type bindingBaseType, String methodName, Type returnType, TypenullableReturnType, Type thisType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 783
       at Starcounter.Binding.BindingBuilder.CreateBooleanPropertyBinding(PropertyDef propertyDef, Type thisType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 277
       at Starcounter.Binding.BindingBuilder.BuildPropertyBindingList(TypeBinding typeBinding, Type type) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 195
       at Starcounter.Binding.BindingBuilder.CreateTypeBinding() in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 151
       at Starcounter.Binding.Bindings.BuildTypeBindingFromTypeDef(TypeDef typeDef)in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\Bindings.cs:line 255
       at Starcounter.Query.Execution.RowTypeBinding.AddTypeBinding(String typeName) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\Execution\Row\RowTypeBinding.cs:line 122
       at Starcounter.Query.Sql.Creator.CreateInitialRowTypeBinding(Term typeListTerm) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\SQL\_Creator.cs:line 392
       at Starcounter.Query.Sql.Creator.CreateRowTypeBinding(Term typeDefTerm, VariableArray varArray) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\SQL\_Creator.cs:line 356
       at Starcounter.Query.QueryPreparation.PrepareOrExecuteQuery[T](String query,Boolean slowSql, Object[] values) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\QueryPreparation.cs:line 73
       at Starcounter.Query.Sql.SqlEnumCache.CacheOrExecuteEnumerator[T](String query, Boolean slowSQL, Object[] values) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Query\SQL\_SqlEnumCache.cs:line 123
       at Starcounter.Db.FullQueryProcess[T](String query, Boolean slowSQL, Object[] values) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Mock\Db.cs:line 126
       at Starcounter.Reload.Unload(String fileName, UInt64 shiftId, Boolean unloadAll) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Mock\Reload.cs:line 80
       at UnloadDatabase.Program.Main(String[] args) in c:\Program Files\Starcounter
    \2.0.2991.3\cli-apps\UnloadDatabase.cs:line 24
    HResult=-2147467261

**Comment by ivarbergman** [comment]
More info:

I created a new Starcounter Application (New project in VS)

    using System;
    using Starcounter;
    
    namespace StarcounterApplication1
    {
        [Database]
        public class MyUser
        {
            public String Name;
            public String Description;
        }
        public class Program
        {
            static void Main()
            {
                Db.Transaction(() =>
                {
                    var user = new MyUser() {Name = "Ivar"};
                });
            }
        }
    }

Using Starcounter Polyjuice Edition I get an Exception. But using the standard Starcounter Edition unload works. Could it be related to some issues in the Concept.Ring-classes?

**Comment by k-rus** [comment]
Thank you for reporting and the details!
The issue seems to be different, but we can keep them in one, since they are both about unload. Yes, I think the problem is related to SO and latest changes in it. I will start investigation on Monday.

**Comment by k-rus** [comment]
> I created a new Starcounter Application (New project in VS)
> 
> using System;
> using Starcounter;
> 
> namespace StarcounterApplication1
> {
>     [Database]
>     public class MyUser
>     {
>         public String Name;
>         public String Description;
>     }
>     public class Program
>     {
>         static void Main()
>         {
>             Db.Transaction(() =>
>             {
>                 var user = new MyUser() {Name = "Ivar"};
>             });
>         }
>     }
> }
> Using Starcounter Polyjuice Edition I get an Exception. But using the standard Starcounter Edition unload works. Could it be related to some issues in the Concept.Ring-classes?

Thank you for the simple reproduction application! I was able to reproduce the problem.

**Comment by k-rus** [comment]
The problem is a protected property `_canBeOrdered` in class `Concepts.Ring2.VendibleLevel`, which is not properly handled by Binder. I re-assign the issue to @per-samuelsson

**Comment by per-samuelsson** [comment]
> I re-assign the issue to @per-samuelsson

Why to me? It's been known since long that non-public data does not work. It did in 1.x, but never in 2.x since the new metadata was introduced.

I fixed this by a really largish redesign in **next**. In **develop** is status quo.

**Comment by k-rus** [comment]
We discussed the issue. The work around is to fix SO that it does not define any private/protected database properties or fields, since it is not handle in the current version of Starcounter. This has to be changed back when new Starcounter version, currently *next*, is in place. I guess @andwah will do work around and then the issue will be placed in backlog to test with *next* version in future.

**Comment by malx122** [comment]
We need a faster solution than that. We really stuck right now and need an update ASAP.

**Comment by per-samuelsson** [comment]
> We need a faster solution than that. We really stuck right now and need an update ASAP.

I think @andwah will push it today. Isn't that fast enough?

**Comment by malx122** [comment]
Ah, I think I might have missunderstand @k-rus message. Waiting on needles (does it work in english?) for a solution :).

**Comment by andwah** [comment]
I have now pushed a fix into Society-Objects repo.
I changed all non public properties to public properties.
(it will be in the next build of the Polyjuice Edition)

**Comment by k-rus** [comment]
My suggestion for a better fix of the problem is in #2508.

**Comment by k-rus** [comment]
The first exception, which is in the head of the issue, is reproduced by simple application and unload. The application code:
```c#
using Starcounter;
using System;

[Database]
public class ReproduceClass {
    static void Main() {
        Db.Transaction(delegate {
            new ReproduceClass();
        });
    }
    private string Name;
}
```

To reproduce:

1. Install Starcounter
2. Run the above application, e.g., `star UnloadReproduce.cs`
3. Do unload, e.g., `staradmin unload --file=here.sql`
4. Unload will fail with the exception:

> Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entrypoint method raised an exception. System.InvalidOperationException: Sequence contains no elements

The suggested fix is #2508.

**Comment by k-rus** [comment]
Another suggested fix to help in the reported situation is #2509.

**Comment by k-rus** [comment]
> Meanwhile..
> I tried "staradmin unload" on a database running the microLauncher app. This also resulted in an exception. (But not exactly the same exception as my first post in this issue.) Is this a separate issue?
> 
> 1 - Fresh install of latest Starcounter Polyjuice version. Starcounter-DevelopNightlyBuilds-2.0.2954
> 2 - Clone https://github.com/Polyjuice/microLauncher
> 3 - Configure Working Directory path to Client-folder 
> 4 - Start the project in VS
> 5 - Access http://localhost:8080/index
> 6 - staradmin unload
> 
> C:\Program Files\Starcounter\2.0.2991.3>staradmin unload
> Unloading Default (unloading ..)
> 
> [18:23:58, Starcounter (default)]
> Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entry point method raised an exception. System.NullReferenceException: Object reference not set to an instance of an object.
>    at Starcounter.Binding.BindingBuilder.VerifyProperty(PropertyInfo propertyInfo, Type returnType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 967
>    at Starcounter.Binding.BindingBuilder.GeneratePropertyBindingDefault(PropertyDef propertyDef, Type bindingBaseType, String methodName, Type returnType, TypenullableReturnType, Type thisType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 783
...

Thank you for the reproduction steps! I was able to reproduce this problem.

**Comment by k-rus** [comment]
> Meanwhile..
> I tried "staradmin unload" on a database running the microLauncher app. This also resulted in an exception. (But not exactly the same exception as my first post in this issue.) Is this a separate issue?
> 
> 1 - Fresh install of latest Starcounter Polyjuice version. Starcounter-DevelopNightlyBuilds-2.0.2954
> 2 - Clone https://github.com/Polyjuice/microLauncher
> 3 - Configure Working Directory path to Client-folder 
> 4 - Start the project in VS
> 5 - Access http://localhost:8080/index
> 6 - staradmin unload
> 
> C:\Program Files\Starcounter\2.0.2991.3>staradmin unload
> Unloading Default (unloading ..)
> 
> [18:23:58, Starcounter (default)]
> Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entry point method raised an exception. System.NullReferenceException: Object reference not set to an instance of an object.
> at Starcounter.Binding.BindingBuilder.VerifyProperty(PropertyInfo propertyInfo, Type returnType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 967
> at Starcounter.Binding.BindingBuilder.GeneratePropertyBindingDefault(PropertyDef propertyDef, Type bindingBaseType, String methodName, Type returnType, TypenullableReturnType, Type thisType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 783
> ...

This is due to the same file `VendibleLevel` is being in projects `Concepts.Ring2` and `Concepts.Ring4`. I have no idea why C# compiler and Starcounter accept this. May be @per-samuelsson can clarify.

**Comment by k-rus** [comment]
> This is due to the same file VendibleLevel is being in projects Concepts.Ring2 and Concepts.Ring4. I have no idea why C# compiler and Starcounter accept this. May be @per-samuelsson can clarify.

I was wrong about it. I was mislead by VS.

**Comment by per-samuelsson** [comment]
This issue is now somewhat resolved, as described in the last comments of #2508. I'll keep it open until we have properly documented it and checked it with @malx122.

**Comment by k-rus** [comment]
> Meanwhile..
> I tried "staradmin unload" on a database running the microLauncher app. This also resulted in an exception. (But not exactly the same exception as my first post in this issue.) Is this a separate issue?
> 
> 1 - Fresh install of latest Starcounter Polyjuice version. Starcounter-DevelopNightlyBuilds-2.0.2954
> 2 - Clone https://github.com/Polyjuice/microLauncher
> 3 - Configure Working Directory path to Client-folder 
> 4 - Start the project in VS
> 5 - Access http://localhost:8080/index
> 6 - staradmin unload
> 
> C:\Program Files\Starcounter\2.0.2991.3>staradmin unload
> Unloading Default (unloading ..)
> 
> [18:23:58, Starcounter (default)]
> Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entry point method raised an exception. System.NullReferenceException: Object reference not set to an instance of an object.
> at Starcounter.Binding.BindingBuilder.VerifyProperty(PropertyInfo propertyInfo, Type returnType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 967
> at Starcounter.Binding.BindingBuilder.GeneratePropertyBindingDefault(PropertyDef propertyDef, Type bindingBaseType, String methodName, Type returnType, TypenullableReturnType, Type thisType) in c:\TeamCity\TeamCity8\buildAgent\work\sc-2.0.2991.3\Level1\src\Starcounter\Binding\BindingBuilder.cs:line 783
> ...
> Thank you for the reproduction steps! I was able to reproduce this problem.

The simple app to reproduce the problem is:
```c#
using System;
using Starcounter;

namespace StarcounterApplication1
{
    [Database]
    public class MyUser
    {
        public String Name;
		protected String _Description;
		[SynonymousTo("_Description")]
        public readonly String Description;
    }
	
    public class Program
    {
        static void Main()
        {
            Db.Transaction(() =>
            {
                var user = new MyUser() {Name = "Ivar"};
            });
        }
    }
}
```
Just start the app in an empty database. I will check how this application behaves in the latest version of Starcounter after @per-samuelsson's fix.

**Comment by k-rus** [comment]
> The simple app to reproduce the problem is:
> 
> using System;
> using Starcounter;
> 
> namespace StarcounterApplication1
> {
>     [Database]
>     public class MyUser
>     {
>         public String Name;
>         protected String _Description;
>         [SynonymousTo("_Description")]
>         public readonly String Description;
>     }
> 
>     public class Program
>     {
>         static void Main()
>         {
>             Db.Transaction(() =>
>             {
>                 var user = new MyUser() {Name = "Ivar"};
>             });
>         }
>     }
> }
> Just start the app in an empty database. I will check how this application behaves in the latest version of Starcounter after @per-samuelsson's fix.

I confirm that it works now. The database is started and unload is successfully executed.
